rpcs3/Emu/Cell/Modules/cellVoice.cpp

```diff
@@ -1,33 +1,143 @@
-#include "stdafx.h"
+﻿#include "stdafx.h"
 #include "Emu/System.h"
 #include "Emu/Cell/PPUModule.h"
+#include "Emu/IdManager.h"
+#include "Emu/Cell/lv2/sys_event.h"
 
 #include "cellVoice.h"
 
 LOG_CHANNEL(cellVoice);
 
+template <>
+void fmt_class_string<CellVoiceError>::format(std::string& out, u64 arg)
+{
+	format_enum(out, arg, [](CellVoiceError value)
+	{
+		switch (value)
+		{
+			STR_CASE(CELL_VOICE_ERROR_ADDRESS_INVALID);
+			STR_CASE(CELL_VOICE_ERROR_ARGUMENT_INVALID);
+			STR_CASE(CELL_VOICE_ERROR_CONTAINER_INVALID);
+			STR_CASE(CELL_VOICE_ERROR_DEVICE_NOT_PRESENT);
+			STR_CASE(CELL_VOICE_ERROR_EVENT_DISPATCH);
+			STR_CASE(CELL_VOICE_ERROR_EVENT_QUEUE);
+			STR_CASE(CELL_VOICE_ERROR_GENERAL);
+			STR_CASE(CELL_VOICE_ERROR_LIBVOICE_INITIALIZED);
+			STR_CASE(CELL_VOICE_ERROR_LIBVOICE_NOT_INIT);
+			STR_CASE(CELL_VOICE_ERROR_NOT_IMPLEMENTED);
+			STR_CASE(CELL_VOICE_ERROR_PORT_INVALID);
+			STR_CASE(CELL_VOICE_ERROR_RESOURCE_INSUFFICIENT);
+			STR_CASE(CELL_VOICE_ERROR_SERVICE_ATTACHED);
+			STR_CASE(CELL_VOICE_ERROR_SERVICE_DETACHED);
+			STR_CASE(CELL_VOICE_ERROR_SERVICE_HANDLE);
+			STR_CASE(CELL_VOICE_ERROR_SERVICE_NOT_FOUND);
+			STR_CASE(CELL_VOICE_ERROR_SHAREDMEMORY);
+			STR_CASE(CELL_VOICE_ERROR_TOPOLOGY);
+		}
+		return unknown;
+	});
+}
 
-s32 cellVoiceConnectIPortToOPort()
+error_code cellVoiceConnectIPortToOPort(u32 ips, u32 ops)
 {
-	UNIMPLEMENTED_FUNC(cellVoice);
+	cellVoice.todo("cellVoiceConnectIPortToOPort(iPort=0x%x, oPort=0x%x)", ips, ops);
+
+	const auto handler = fxm::get<voice_t>();
+	if (!handler)
+	{
+		return CELL_VOICE_ERROR_LIBVOICE_NOT_INIT;
+	}
+
 	return CELL_OK;
 }
 
-s32 cellVoiceCreateNotifyEventQueue()
+error_code cellVoiceCreateNotifyEventQueue(vm::ptr<u32> id, vm::ptr<u64> key)
 {
-	UNIMPLEMENTED_FUNC(cellVoice);
-	return CELL_OK;
+	cellVoice.todo("cellVoiceCreateNotifyEventQueue(id=*0x%x, key=*0x%x)", id, key);
+
+	const auto handler = fxm::get<voice_t>();
+	if (!handler)
+	{
+		return CELL_VOICE_ERROR_LIBVOICE_NOT_INIT;
+	}
+
+	// axed from cellAudio
+	vm::var<sys_event_queue_attribute_t> attr;
+	attr->protocol = SYS_SYNC_FIFO;
+	attr->type = SYS_PPU_QUEUE;
+	attr->name_u64 = 0;
+
+	for (u64 i = 0; i < 10; i++)
+	{
+		// Create an event queue "bruteforcing" an available key
+		const u64 key_value = 0x80004d494f323285ull + i;
+		if (const s32 res = sys_event_queue_create(id, attr, key_value, 32))
+		{
+			if (res != CELL_EEXIST)
+			{
+				return res;
+			}
+		}
+		else
+		{
+			*key = key_value;
+			return CELL_OK;
+		}
+	}
+
+	return CELL_VOICE_ERROR_EVENT_QUEUE;
 }
 
-s32 cellVoiceCreatePort()
+error_code cellVoiceCreatePort(vm::ptr<u32> portId, vm::cptr<CellVoicePortParam> pArg)
 {
-	UNIMPLEMENTED_FUNC(cellVoice);
+	cellVoice.todo("cellVoiceCreatePort(portId=*0x%x, pArg=*0x%x)", portId, pArg);
+
+	if (!pArg)
+	{
+		return CELL_VOICE_ERROR_ARGUMENT_INVALID;
```

Comment:
The first check in this function should be for LIBVOICE_NOT_INIT, this is usually the same for all of the libraries.